org.metaborg.spoofax.core/src/main/java/org/metaborg/spoofax/core/dynamicclassloading/DynamicClassLoadingService.java

```diff
@@ -0,0 +1,138 @@
+package org.metaborg.spoofax.core.dynamicclassloading;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.Set;
+
+import org.apache.commons.vfs2.FileObject;
+import org.metaborg.core.MetaborgException;
+import org.metaborg.core.language.ILanguageCache;
+import org.metaborg.core.language.ILanguageComponent;
+import org.metaborg.core.language.ILanguageImpl;
+import org.metaborg.core.resource.IResourceService;
+import org.metaborg.util.log.ILogger;
+import org.metaborg.util.log.LoggerUtils;
+
+import com.google.common.collect.Maps;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+
+public class DynamicClassLoadingService implements IDynamicClassLoadingService, ILanguageCache {
+    private static final ILogger logger = LoggerUtils.logger(DynamicClassLoadingService.class);
+
+    private final IResourceService resourceService;
+    private final Set<ClassLoader> additionalClassLoaders;
+
+    private final Map<ILanguageComponent, ClassLoader> classLoaderCache = Maps.newHashMap();
+    @SuppressWarnings("rawtypes")
+    private final Map<ILanguageComponent, Map<Class, ServiceLoader>> serviceLoaderCache = Maps.newHashMap();
+    private final Injector injector;
+
+
+    @Inject public DynamicClassLoadingService(IResourceService resourceService,
+        Set<ClassLoader> additionalClassLoaders, Injector injector) {
+        this.resourceService = resourceService;
+        this.additionalClassLoaders = additionalClassLoaders;
+        this.injector = injector;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> T loadClass(ILanguageComponent component, String className, Class<T> expectedType) throws MetaborgException {
+        final ClassLoader classLoader = classLoader(component);
+        Class<?> theClass;
+        try {
+            logger.trace("Loading outliner class");
+            theClass = classLoader.loadClass(className);
+        } catch (ClassNotFoundException e) {
+            throw new MetaborgException("Given class was not found: " + className, e);
+        }
+        T instance;
+        try {
+            logger.trace("Instantiating outliner class");
+            instance = (T) theClass.newInstance();
+            injector.injectMembers(instance);
+        } catch (InstantiationException e) {
+            throw new MetaborgException("Given class was not instantiable", e);
```

Comment:
Could you include the class names in these exceptions? That would make it easier to debug if it fails.